### DISCUSSION TITLE
Custom RT1064 can not be flashed with Jlink on Zephyr

### DISCUSSION THREAD

**Discussion by mariopaja** [zephyrproject-rtos/zephyr#74089] [Q&A] [answered]
We have a custom board RT1064 which we are not able to flash with jlink on Zephyr v3.6.0. 
However flashing a bare metal sample with MCUXpresso worked without any problem.

At the end we get the following error: `****** Error: Could not start CPU core. (ErrorCode: -1)`

Any help would be appreciated :)

```
runners.jlink: /opt/SEGGER/JLink_V796h/JLinkExe -nogui 1 -if swd -speed auto -device MIMXRT1064 -CommanderScript /tmp/tmpdfrmhguojlink/runner.jlink -nogui 1
SEGGER J-Link Commander V7.96h (Compiled May 15 2024 15:40:18)
DLL version V7.96h, compiled May 15 2024 15:39:51


J-Link Command File read successfully.
Processing script file...
J-Link>ExitOnError 1
J-Link Commander will now exit on Error
J-Link>r
J-Link connection not established yet but required for command.
Connecting to J-Link via USB...O.K.
Firmware: J-Link V11 compiled Mar 14 2024 13:16:08
Hardware version: V11.00
J-Link uptime (since boot): 0d 00h 42m 02s
S/N: 601006160
License(s): RDI, FlashBP, FlashDL, JFlash, GDB
USB speed mode: High speed (480 MBit/s)
VTref=3.328V
Target connection not established yet but required for command.
Device "MIMXRT1064XXX6A" selected.


Connecting to target via SWD
InitTarget() start
InitTarget() end - Took 11.2ms
Found SW-DP with ID 0x0BD11477
Failed to power up DAP
InitTarget() start
InitTarget() end - Took 5.66ms
Found SW-DP with ID 0x0BD11477
DPIDR: 0x0BD11477
CoreSight SoC-400 or earlier
Scanning AP map to find all available APs
AP[1]: Stopped AP scan as end of AP map has been reached
AP[0]: AHB-AP (IDR: 0x04770041)
Iterating through AP map to find AHB-AP to use
AP[0]: Core found
AP[0]: AHB-AP ROM base: 0xE00FD000
CPUID register: 0x411FC271. Implementer code: 0x41 (ARM)
Cache: L1 I/D-cache present
Found Cortex-M7 r1p1, Little endian.
FPUnit: 8 code (BP) slots and 0 literal slots
CoreSight components:
ROMTbl[0] @ E00FD000
[0][0]: E00FE000 CID B105100D PID 000BB4C8 ROM Table
ROMTbl[1] @ E00FE000
[1][0]: E00FF000 CID B105100D PID 000BB4C7 ROM Table
ROMTbl[2] @ E00FF000
[2][0]: E000E000 CID B105E00D PID 000BB00C SCS-M7
[2][1]: E0001000 CID B105E00D PID 000BB002 DWT
[2][2]: E0002000 CID B105E00D PID 000BB00E FPB-M7
[2][3]: E0000000 CID B105E00D PID 000BB001 ITM
[1][1]: E0041000 CID B105900D PID 001BB975 ETM-M7
[1][2]: E0042000 CID B105900D PID 004BB906 CTI
[0][1]: E0040000 CID B105900D PID 000BB9A9 TPIU-M7
[0][2]: E0043000 CID B105F00D PID 001BB101 TSG
I-Cache L1: 32 KB, 512 Sets, 32 Bytes/Line, 2-Way
D-Cache L1: 32 KB, 256 Sets, 32 Bytes/Line, 4-Way
Memory zones:
  Zone: "Default" Description: Default access mode
Cortex-M7 identified.
Reset delay: 0 ms
Reset type NORMAL: Resets core & peripherals via SYSRESETREQ & VECTRESET bit.
ResetTarget() start
ResetTarget() end - Took 159ms
AfterResetTarget() start
AfterResetTarget() end - Took 5.70ms
J-Link>loadfile "/home/mario/Repos/workspace/zephyr/samples/hello_world/build/zephyr/zephyr.bin" 0x70000000
'loadfile': Performing implicit reset & halt of MCU.
ResetTarget() start
ResetTarget() end - Took 145ms
AfterResetTarget() start
AfterResetTarget() end - Took 4.07ms
Downloading file [/home/mario/Repos/workspace/zephyr/samples/hello_world/build/zephyr/zephyr.bin]...
Comparing flash   [100%] Done.
Erasing flash     [100%] Done.
Programming flash [100%] Done.
J-Link: Flash download: Bank 0 @ 0x70000000: 1 range affected (65536 bytes)
J-Link: Flash download: Total: 0.705s (Prepare: 0.072s, Compare: 0.084s, Erase: 0.256s, Program & Verify: 0.255s, Restore: 0.037s)
J-Link: Flash download: Program & Verify speed: 251 KB/s
O.K.
J-Link>g
Memory map 'after startup completion point' is active

****** Error: Could not start CPU core. (ErrorCode: -1)
J-Link>writeDP 1 0
Writing failed.
J-Link>readDP 1
Reading failed.
J-Link>q

Script processing completed.
```

**Comment by DerekSnell** [comment]
Hi @mariopaja ,
I would try a few experiments to help narrow down the issue.  For example, you said the MCUXpresso IDE can program a baremetal image.  Can you try using MCUXpresso IDE and its GUI flash programmer to program your Zephyr image binary at /home/mario/Repos/workspace/zephyr/samples/hello_world/build/zephyr/zephyr.bin?  This will help confirm if the issue is in the image or with the tools.  Or another option is to use the JLink CLI tool with the loadfile command to program the baremetal image.

Also, is the FCB and IVT required to boot from the FlexSPI flash included in your Zephyr image?  You can view these boot structures are the start of the image in the two binary files and see if they are different.  If they are missing from your Zephyr image, then perhaps your custom board is not including them.

Another thought, since you are booting from the integrated flash on the FlexSPI2, this is the same as our EVK.  You could try building a Zephyr app for the mimxrt1064_evk board, and try loading that on your board.  I expect the ROM would be able to boot it up, even if the app does not run correctly due to board differences.  So I would expect the Jlink to be able to flash that image successfully to your board.  If so, that would also point to a difference with your custom board config.

Let us know what you find.

**Reply by mariopaja** [reply]
@DerekSnell Thank you for your tips and sorry for my late reply, only now I could come back into this topic.

> I would try a few experiments to help narrow down the issue. For example, you said the MCUXpresso IDE can program a baremetal image. Can you try using MCUXpresso IDE and its GUI flash programmer to program your Zephyr image binary at /home/mario/Repos/workspace/zephyr/samples/hello_world/build/zephyr/zephyr.bin? This will help confirm if the issue is in the image or with the tools. Or another option is to use the JLink CLI tool with the loadfile command to program the baremetal image.

I could not flash the zephyr image with the MCUXpresso GUI Flash, ar at least I had the same results as before :/
Jlink could flash the bare metal image

> Another thought, since you are booting from the integrated flash on the FlexSPI2, this is the same as our EVK. You could try building a Zephyr app for the mimxrt1064_evk board, and try loading that on your board. I expect the ROM would be able to boot it up, even if the app does not run correctly due to board differences. So I would expect the Jlink to be able to flash that image successfully to your board. If so, that would also point to a difference with your custom board config.

Flashing our custom board with the mimxrt1064_evk image did not fix the issue :/
Simple hello world application should work becuase we use the same uart as the EVK

> Also, is the FCB and IVT required to boot from the FlexSPI flash included in your Zephyr image? You can view these boot structures are the start of the image in the two binary files and see if they are different. If they are missing from your Zephyr image, then perhaps your custom board is not including them.

Left - Bare metal
Middle - custom board (zephyr)
Right - EVK (zepphyr)

![Screenshot 2024-06-24 103707](https://github.com/zephyrproject-rtos/zephyr/assets/28357816/2d6c9726-7422-48f7-b0e4-e632fad81553)
![Screenshot 2024-06-24 103915](https://github.com/zephyrproject-rtos/zephyr/assets/28357816/d4f7d016-ec9f-415d-b113-a608d213deb2)

On MCUXpresso on sdram_ini_dcd i found several differences compared to the mimxrt1064 and I already changed it to a local file. However, it did not change the behaviour

**Comment by DerekSnell** [comment, marked as answer]
Hi @mariopaja ,
The baremetal image that works does not include the DCD, which configures for SDRAM.  Does your board have SDRAM?  If so, is it the same used on the EVK?  The default Zephyr image will place data in SDRAM, and try to access it early in the startup code.  I would move the data `zephyr,sram` node to DTCM to rule out issues with the SDRAM.

And your flash dumps show the flash is programmed with the Zephyr images.  I suspect the device is booting, but runs into an issue early in the startup code, and faults or runs away.  Can you use a debugger and halt in the reset entry function?  That would confirm it is booting.  And then you can step through the code and see how far it gets.

Let us know what you find.

**Reply by mariopaja** [reply]
@DerekSnell  Thank you it seems the issue is related to the SDRAM! 


> Does your board have SDRAM? If so, is it the same used on the EVK?

Yes the board has SDRAM and yes it is the same as the evk.

> I would move the data `zephyr,sram` node to DTCM to rule out issues with the SDRAM.

I just did it and the mcu booted without any problem. 


I also tested the SDRAM with the bare metal SEMC SDRAM Example. The 32-bit and 16-bit  test fail. Only the 8bit test passed :/

```
 SEMC SDRAM Example Start!

 SEMC SDRAM Memory 32 bit Write Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM Read 32 bit Data Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM 32 bit Data Write and Read Compare Start!

 SEMC SDRAM 32 bit Data Write and Read Compare Failed!

 SEMC SDRAM Memory 16 bit Write Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM Read 16 bit Data Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM 16 bit Data Write and Read Compare Start!

 SEMC SDRAM 16 bit Data Write and Read Compare Failed!

 SEMC SDRAM Memory 8 bit Write Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM Read 8 bit Data Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM 8 bit Data Write and Read Compare Start!

 SEMC SDRAM 8 bit Data Write and Read Compare Succeed!

 SEMC SDRAM Example End. 

```

However when I disable EMC_39 only the 32bit test fails.
For context we are using EMC_39 as a gpio to control an LED, although normally it is used as SEMC_DQS.
At this point I would consider this a HW related problem and I will have a deeper look into it with the HW designer.
Thank you once again!

```
 SEMC SDRAM Example Start!

 SEMC SDRAM Memory 32 bit Write Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM Read 32 bit Data Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM 32 bit Data Write and Read Compare Start!

 SEMC SDRAM 32 bit Data Write and Read Compare Failed!

 SEMC SDRAM Memory 16 bit Write Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM Read 16 bit Data Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM 16 bit Data Write and Read Compare Start!

 SEMC SDRAM 16 bit Data Write and Read Compare Succeed!

 SEMC SDRAM Memory 8 bit Write Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM Read 8 bit Data Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM 8 bit Data Write and Read Compare Start!

 SEMC SDRAM 8 bit Data Write and Read Compare Succeed!

 SEMC SDRAM Example End.

```

**Reply by mariopaja** [reply]
@DerekSnell The SDRAM itself works but it seems like it takes too long for the SDRAM to become available for write/read

2 consecutive 32bit write/read  - bare metal sample
```
 SEMC SDRAM Example Start!

 SEMC SDRAM Memory 32 bit Write Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM Read 32 bit Data Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM 32 bit Data Write and Read Compare Start!

 SEMC SDRAM 32 bit Data Write and Read Compare Failed!

 SEMC SDRAM Memory 32 bit Write Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM Read 32 bit Data Start, Start Address 0x80000000, Data Length 4096 !

 SEMC SDRAM 32 bit Data Write and Read Compare Start!

 SEMC SDRAM 32 bit Data Write and Read Compare Succeed!

 SEMC SDRAM Example End.
```

A printout of write/read data:

 `sdram_writeBuffer[datalen]= x020004018, sdram_readBuffer[datalen]= x0fff`
